opte/src/engine/port.rs

```diff
@@ -693,6 +719,9 @@ impl Port {
             });
         }
 
+        // Perhaps this is silly, but I want to make sure the state
+        // lock is held for the execution of this method.
+        drop(state);
```

Comment:
Actually, I'm looking at the lock in more detail now. It seems like it protects just the `PortState`, but this clearly implies it protects more than that, is that right? If so, we may want to put all the data that needs synchronization behind the lock. If that's a lot of fields, on common pattern is to put everything that needs to be locked into its own "inner" struct, and then storing that in the mutex.

Reply:
Yea, so this kind of happened for a bunch of different reasons. Part of it has to do with my influence from illumos/mac/C and how we typically structure protection of resources. We usually have multiple locks to allow a semi-fine-grained approach in order to minimize contention. Another typical pattern is to prevent the admin path from contending with the data path. In the case of mac this is done via the "mac perimeter" which makes sure all admin commands execute sequentially but does not block any data path traffic. However, occasionally an admin command needs to coordinate-with or freeze the datapath, and mac does that with a system of per-CPU locks and reference counts. I'm not quite sure how you do something like that in Rust, as you aren't really partitioning the ownership but rather the locks that say "yes, you can access this". Though, you could use the flow id/5-tuple to truly partition ownership, and then I think this maps better with Rust.

All that said, you are totally right to call this out, as it's basically me trying to map my C/illumos like idioms into Rust. I agree that in Rust the idea is for the mutex to represent ownership of a thing at runtime, allowing one to publish something to multiple consumers without fear of corruption or races. To me it was important that I kept some form of "performance" by using the design I currently have, but now I'm realizing 1) it assumes the consumer of the generic OPTE is mac (as in it assumes that mc_start/mc_stop is how a port is started/stopped) and 2) you could theoretically have a race where the port is paused and then one more packet is processed. That could be detrimental to the save routine (which isn't written yet but needs to be) as it iterates the layer/UFT state, as it could interleave its reading of state with the packet's mutation of state.

This is making me think that the first thing to do is make things obviously correct but using a BFM (big fucking mutex), and thus forcing admin and data to contend with each other. However, when this becomes a hinderance on perf we could transition to something more like what mac does and partition ownership on some mechanism. For example, I already want to split out the UFT into TCP, UDP, and Other protocols. So that could be a natural place to split ownership of state. Then, when an admin command needs to coordinate with the datapath, it could use a similar mechanism to mac to quiesce the data.

Although, if we make it a requirement that opte is always executed in the context of a mac provider (like xde), then perhaps we take advantage of that fact? We would make sure the mac perim is used for all admin commands (by making sure the DLD entrypoint grabs the perim...assuming that's possible), and then on the datapath just assume the port's state cannot change while a packet is being processed (because this is exactly what mac promises to a provider). You still need the various fine-grained locks currently in opte as you can have concurrent data threads, but you could get rid of the port state lock as it can only change when data is quiesced and there is no traffic flowing through the port.

I need to think through that more and maybe talk with Robert, but certainly this has raised the need to write up an issue. For the purposes of this PR I want to leave it as-is, as it works fine for our purposes at the moment.

Reply:
Thanks for the details, that's very helpful context.

It's true that Rust's `Mutex` owns the data it locks, which is usually presented as a good thing since one can't inadvertently access a field that is protected by the lock. You just can't get that field without acquiring the lock. I don't think that precludes fine-grained locking, since we can group fields that all require the same synchronization (e.g., admin vs data-path) into separate `Mutex`es. I don't know enough about MAC to understand how that interacts.